Add Makefile, configuration files and scripts (#13)
unikraft/app-nginx: README.md

  * The networking stack library: [`lib-lwip`](https://github.com/unikraft/lib-lwip)

Follow the steps below for the setup:

  1. First clone the [`app-nginx` repository](https://github.com/unikraft/app-nginx) repository in the `nginx/` directory:

     ```console
     $ git clone https://github.com/unikraft/app-nginx nginx
     ```

     Enter the `nginx/` directory:

     ```console
     $ cd nginx/

     $ ls
     fs0/  kraft.yaml  Makefile  Makefile.uk.default  README.md  run-aarch64*  run-x86_64*
     ```

  1. While inside the `nginx/` directory, create the `.unikraft/` directory:

     ```console
     $ mkdir .unikraft
     ```

     Enter the `.unikraft/` directory:

     ```console
     $ cd .unikraft/
     ```

  1. While inside the `.unikraft` directory, clone the [`unikraft` repository]():

     ```console
     $ git clone https://github.com/unikraft/unikraft unikraft
     ```

  1. While inside the `.unikraft/` directory, create the `libs/` directory:

     ```console
     $ mkdir libs
     ```

  1. While inside the `.unikraft/` directory, clone the library repositories in the `libs/` directory:

>      ```console

Reviewer: ```suggestion
     ```console
     $ cd libs/
```

     $ git clone https://github.com/unikraft/lib-nginx nginx

     $ git clone https://github.com/unikraft/lib-musl musl

     $ git clone https://github.com/unikraft/lib-lwip lwip
     ```

  1. Get back to the application directory:

     ```console

>      $ cd ../

Reviewer: ```suggestion
     $ cd ../../
```

     ```

     Use the `tree` command to inspect the contents of the `.unikraft/` directory.
     It should print something like this:

     ```console

>      $ tree --charset=ascii -F -L 2 .unikraft/

Reviewer: ```suggestion
     $ tree -F -L 2 .unikraft/
```

     .unikraft/
     |-- libs/
     |   |-- lwip/
     |   |-- musl/
     |   `-- nginx/
     `-- unikraft/
         |-- arch/
         |-- Config.uk
         |-- CONTRIBUTING.md
         |-- COPYING.md
         |-- include/
         |-- lib/
         |-- Makefile
         |-- Makefile.uk
         |-- plat/
         |-- README.md
         |-- support/
         `-- version.mk

     10 directories, 7 files
     ```

## Configure

Configuring, building and running a Unikraft application depends on our choice of platform and architecture.
Currently supported platforms are QEMU (KVM), Xen and linuxu.
QEMU (KVM) is known to be working, so we focus on that.

Supported architectures are x86_64 and AArch64.

Use the corresponding the configuration files (`config-...`), according to your choice of platform and architecture.

### QEMU x86_64

Use the `config-qemu-x86_64` configuration file together with `make defconfig` to create the configuration file:

```console

> $ UK_DEFCONFIG=$(pwd)/config-qemu-x84_64 make defconfig

Reviewer: ```suggestion
$ UK_DEFCONFIG=$(pwd)/config-qemu-x86_64 make defconfig
```

The unikernel output image, together with intermediary build files, are stored in the `build/` directory.

### Clean Up

Before starting a build on a different platform or architecture, you must clean up the build output.
This may also be required in case of a new configuration.

Cleaning up is done with 3 possible commands:

* `make clean`: cleans all actual build output files (binary files, including the unikernel image)
* `make properclean`: removes the entire `build/` directory
* `make distclean`: removes the entire `build/` directory **and** the `.config` file

Typically, you would use `make properclean` to remove all build artifacts, but keep the configuration file.

### QEMU x86_64

Building for QEMU x86_64 assumes you did the QEMU x86_64 configuration step above.
Build the Unikraft Nginx image for QEMU AArch64 by using the command below:

```console
$ make -j $(nproc)
[...]
  LD      nginx_qemu-x86_64.dbg
  UKBI    nginx_qemu-x86_64.dbg.bootinfo
  SCSTRIP nginx_qemu-x86_64
  GZ      nginx_qemu-x86_64.gz
make[1]: Leaving directory '/media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/solo/nginx/.unikraft/unikraft'
```

> At the end of the build command, the `app-nginx_qemu-x86_64` unikernel image is generated.

Reviewer: ```suggestion
At the end of the build command, the `nginx_qemu-x86_64` unikernel image is generated.
```

This image is to be used in the run step.

### QEMU AArch64

Building for QEMU AArch64 requires a cross-compile toolchain.
The instructions for installing the cross-compile toolchain are part of the [`Requirements` section](#requirements).

If you had configured and build a unikernel image for another platform or architecture (such as x86_64) before, then:

1. Do a cleanup step with `make properclean`.

1. Configure for QEMU AAarch64, as shown above.

1. Follow the instructions below to build for QEMU AArch64.

Build the Unikraft Nginx image for QEMU AArch64 by using the command below.
Replace `<path/to/cross/compiler>` with the path to the cross compile toolchain (as part of the `CROSS_COMPILE` build variable):

```console
$ make CROSS_COMPILE=<path/to/cross/compiler> -j $(nproc)
```

A sample command is:

```
$ make CROSS_COMPILE=~/packages/gcc-arm-11.2-2022.02-x86_64-aarch64-none-linux-gnu/bin/aarch64-none-linux-gnu- -j $(nproc)
[...]

>   LD      app-nginx_qemu-arm64.dbg
>   UKBI    app-nginx_qemu-arm64.dbg.bootinfo
>   SCSTRIP app-nginx_qemu-arm64
>   GZ      app-nginx_qemu-arm64.gz

Reviewer: ```suggestion
  LD      nginx_qemu-arm64.dbg
  UKBI    nginx_qemu-arm64.dbg.bootinfo
  SCSTRIP nginx_qemu-arm64
  GZ      nginx_qemu-arm64.gz
```

make[1]: Leaving directory '/media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/solo/app-nginx/.unikraft/unikraft
```

Be sure to pass a string ending in `-`;
that is, DO NOT fill the command name: `aarch64-none-linux-gnu-gcc`;
simply end it with `aarch64-none-linux-gnu-`.

> Similarly to x86_64, at the end of the build command, the `app-nginx_qemu-arm64` unikernel image is generated.

Reviewer: ```suggestion
Similarly to x86_64, at the end of the build command, the `nginx_qemu-arm64` unikernel image is generated.
```

```console
$ UK_DEFCONFIG=$(pwd)/config-qemu-aarch64 make defconfig
```

Similar to the x86_64 configuration, this results in the creation of the `.config` file that will be used in the build step.

## Build

Building uses as input the `.config` file from above, and results in a unikernel image as output.
The unikernel output image, together with intermediary build files, are stored in the `build/` directory.

### Clean Up

Before starting a build on a different platform or architecture, you must clean up the build output.
This may also be required in case of a new configuration.

Cleaning up is done with 3 possible commands:

* `make clean`: cleans all actual build output files (binary files, including the unikernel image)
* `make properclean`: removes the entire `build/` directory
* `make distclean`: removes the entire `build/` directory **and** the `.config` file

Typically, you would use `make properclean` to remove all build artifacts, but keep the configuration file.

### QEMU x86_64

Building for QEMU x86_64 assumes you did the QEMU x86_64 configuration step above.

> Build the Unikraft Nginx image for QEMU AArch64 by using the command below:

Reviewer: ```suggestion
Build the Unikraft Nginx image for QEMU x86_64 by using the command below:
```
